File: hackerrank/taleOfTwoStacks.js
// https://www.hackerrank.com/challenges/ctci-queue-using-two-stacks/problem?h_l=interview&playlist_slugs%5B%5D=interview-preparation-kit&playlist_slugs%5B%5D=stacks-queues

function processData(input) {
  const inputs = input.split("\n");
  const inQueue = [];
  const outQueue = [];

  function inQueueToOutQueue() {
    if (outQueue.length === 0) {
      while (inQueue.length !== 0) {
        outQueue.push(inQueue.pop());
      }
    }
  }

  for (let i = 1; i <= inputs[0]; i++) {
    let query = inputs[i].split(" ");
    switch (query[0]) {
      case "1":
        inQueue.push(query[1]);
        break;
      case "2":
        inQueueToOutQueue();
        outQueue.pop();
        break;
      case "3":
        inQueueToOutQueue();
        console.log(outQueue[outQueue.length - 1]);
        break;
    }
  }
}

console.log(processData("10\n1 42\n2\n1 14\n3\n1 28\n3\n1 60\n1 78\n2\n2")); // Should return 14 14

File: hackerrank/Python/taleOfTwoStacks.py
# https://www.hackerrank.com/challenges/ctci-queue-using-two-stacks/problem?h_l=interview&playlist_slugs%5B%5D=interview-preparation-kit&playlist_slugs%5B%5D=stacks-queues


class MyQueue(object):
    def __init__(self):
        self.inQueue = []
        self.outQueue = []

    def peek(self):
        self.inQueueToOutQueue()
        return self.outQueue[-1]

    def pop(self):
        self.inQueueToOutQueue()
        return self.outQueue.pop()

    def put(self, value):
        self.inQueue.append(value)

    def inQueueToOutQueue(self):
        if len(self.outQueue) == 0:
            while len(self.inQueue) > 0:
                self.outQueue.append(self.inQueue.pop())


queue = MyQueue()
t = int(input())
for line in range(t):
    values = map(int, input().split())
    values = list(values)
    if values[0] == 1:
        queue.put(values[1])
    elif values[0] == 2:
        queue.pop()
    else:
        print(queue.peek())
